dt_parse_iso.c

```diff
@@ -496,69 +570,89 @@ dt_parse_iso_date(const char *str, size_t len, dt_t *dtp) {
         case 0: /* 2012-Q4-85 | 2012-W52-1 */
             break;
         case 2: /* 2012-12-24 */
-            if (p[7] != '-' || count_digits(p, 8, len) != 2)
+            if (p[3] != '-' || count_digits(p, 4, len) != 2)
                 return 0;
-            x = parse_number(p, 5, 2);
-            d = parse_number(p, 8, 2);
-            n = 10;
+            x = parse_number(p, 1, 2);
+            d = parse_number(p, 4, 2);
+            p += 6;
             goto ymd;
         case 3: /* 2012-359 */
-            d = parse_number(p, 5, 3);
-            n = 8;
+            d = parse_number(p, 1, 3);
+            p += 4;
             goto yd;
         default:
             return 0;
     }
 
-    if (len < 10)
+    if (len < (head_n + 6))
         return 0;
 
-    n = count_digits(p, 6, len);
-    switch (p[5]) {
+    n = count_digits(p, 2, len);
+    switch (p[1]) {
 #ifndef DT_PARSE_ISO_STRICT
         case 'Q': /* 2012-Q4-85 */
-            if (n != 1 || p[7] != '-' || count_digits(p, 8, len) != 2)
+            if (n != 1 || p[3] != '-' || count_digits(p, 4, len) != 2)
                 return 0;
-            x = parse_number(p, 6, 1);
-            d = parse_number(p, 8, 2);
-            n = 10;
+            x = parse_number(p, 2, 1);
+            d = parse_number(p, 4, 2);
+            p += 6;
             goto yqd;
 #endif
         case 'W': /* 2012-W52-1 */
-            if (n != 2 || p[8] != '-' || count_digits(p, 9, len) != 1)
+            if (n != 2 || p[4] != '-' || count_digits(p, 5, len) != 1)
                 return 0;
-            x = parse_number(p, 6, 2);
-            d = parse_number(p, 9, 1);
-            n = 10;
+            x = parse_number(p, 2, 2);
+            d = parse_number(p, 5, 1);
+            p += 6;;
             goto ywd;
         default:
             return 0;
     }
 
   yd:
+#ifdef DT_PARSE_ISO_TNT
+    if (!dt_from_yd_checked(sign * y, d, &dt))
+        return 0;
+#else
     if (!dt_valid_yd(y, d))
```

Comment:
We needed to add boundary checks for year ranges we have calculated in `dt_from_*`. But there was no chance to raise error from `dt_from_*` - validation supposed to be done in `dt_valid_*`. So we have merged them together into`dt_from_*_checked` set of functions, and there, once we have calculated `dt` value we could check it for validity and return error if there is overflow detected.